### PR TITLE
Initialize support app export implementation

### PR DESCRIPTION
**Pull request by medamineziraoui**
## Why?
We will add feature top export application from v2 platform to v2 platform.

## Links / Ressources
https://github.com/saagie/gradle-saagie-dataops-plugin/issues/354

## Actual status / Investigations 

Query to get detail of one application
```
labWebApp($id: UUID!) { labWebApp(id: $id) {    id    name    description    creationDate    isDeletable   storageSizeInMB    instances(limit: 1, checkInPipelineInstance: false) {      id     status     statusDetails     startTime      endTime      version {        number        __typename      }      __typename    }    versions {     number      creator      creationDate      number      isCurrent      releaseNote      dockerInfo {        image        dockerCredentialsId       __typename      }      exposedPorts {        name        port        isAuthenticationRequired       isRewriteUrl        basePathVariableName        __typename      }      storagePaths     __typename    }    alerting {      emails      statusList      loginEmails {        login        email       __typename      }      __typename    }    technology {     id      __typename    }    __typename  }}
```
Query   to get all application from the platform
```
query labWebAppsQuery($projectId: UUID!) {  labWebApps(projectId: $projectId) {  id   name  }}
```

Query   to get technology list to be used by the application
```
{
  "operationName": "repositoriesQuery",
  "variables": {},
  "query": "query repositoriesQuery {  repositories {    id    name    technologies {      ... on AppTechnology {        id        label        description        icon        backgroundColor        available        customFlags        __typename      }      __typename    }    __typename  }}"
}
```

Query   to add version to an existing application.
```
{
  "operationName": "addAppVersionMutation",
  "variables": {
    "appId": "35281f01-d25b-4eeb-8a90-8606f298e51c",
    "appVersion": {
      "dockerInfo": {
        "image": "saagie/grafana:7.1.5"
      },
      "exposedPorts": [
        {
          "basePathVariableName": "SAAGIE_BASE_PATH",
          "isRewriteUrl": false,
          "isAuthenticationRequired": false,
          "port": 80,
          "name": "Grafana"
        }
      ],
      "storagePaths": [
        "/opt/grafana",
        "/opt/plugins"
      ],
      "releaseNote": "",
      "resources": {
        "cpu": 0.3,
        "disk": 512,
        "memory": 512
      }
    }
  },
  "query": "mutation addAppVersionMutation($appId: UUID!, $appVersion: JobVersionInput!) {  addJobVersion(jobId: $appId, jobVersion: $appVersion) {    number    __typename  }}"
}
```
## Entites: 
- [ ] `ExportApp` will contain
```
@TypeChecked
class ExportApp implements IExists {
    AppDTO appDTO = new AppDTO()
    AppVersionDTO AppVersionDTO = new AppVersionDTO()
    ArrayList<AppVersionDTO> versions = new ArrayList<AppVersionDTO>()
@Override
    boolean exists() {
        return appDTO.exists() && AppVersionDTO.exists()
    }

    void setAppFromApiResult(appDetailResult) {
        AppDTO.setAppFromApiResult(appDetailResult)
    }

    void setAppVersionFromApiResult(version) {
       appVersionDTO.setAppVersionFromApiResult(version)
    }
}

    void addAppVersionFromV2ApiResult(version) {
        AppVersionDTO appVersionDTOElement = new AppVersionDTO()
       appVersionDTOElement.setAppVersionFromApiResult(version)
        versions.add(appVersionDTOElement)
    }
```
- [ ] `AppDTO` will contain
```
class AppDTO extends JobDTO {
  int storageSizeInMB;
  void setAppFromApiResult(appDetailResult) {
     this.storageSizeInMB = appDetailResult.storageSizeInMB;
     this.setJobFromApiResult(appDetailResult);
  }
}
```
- [ ] `AppVersionDTO` will contain
```
class AppVersionDTO implements IExists {
    List<String> storagePaths = [] //
    String releaseNote //
    int number //
    List<ExposedPort> exposedPorts = [] //
    DockerInfos dockerInfo = new DockerInfos() //
    Resources resources = new Resources() //

    @Override
        boolean exists() {
            return dockerInfo ||
                releaseNote ||
                resources
        }
     void setAppVersionFromApiResult(version) {

            if (version.dockerInfo && version.dockerInfo.dockerCredentialsId) {
                dockerInfo.dockerCredentialsId = version.dockerInfo.dockerCredentialsId
            }

            if (version.dockerInfo && version.dockerInfo.image) {
                dockerInfo.image = version.dockerInfo.image
            }

            releaseNote = version.releaseNote
            storagePaths = version.storagePaths

            if (version.number) {
                number = version.number
            }

            if (version.exposedPorts) {
                exposedPorts = version.exposedPorts
            }

            if (version.resources) {
                resources = version.resources
            }
        }
}

```
- [ ] `AppMapper` will contain
```
package io.saagie.plugin.dataops.models

class AppMapper {
    App job = new App()
    AppVersion jobVersion = new AppVersion()

    public AppMapper() {
        this.AppVersion.resources {
            disk = 512
            memory = 512
            cpu = 0.3
        }
    }

    Object app(Closure closure) {
        app.with(closure)
    }

    Object appVersion(Closure closure) {
        appVersion.with(closure)
    }

    def static Map mapAppAndAppVersionWithoutMail(App app, AppVersion appVersion, String projectId) {
        if (!app) {
            return null
        }
        def newAppConfig = mapAppWithoutMail(app, projectId)
        def mappedAppVersion = appVersion.toMap()


        if (appVersion) {
            newAppConfig.put('jobVersion', [
                releaseNote   : mappedAppVersion.releaseNote
            ])

            if (mappedAppVersion.dockerInfo?.image) {
                newAppConfig.jobVersion.put('dockerInfo', [
                    image              : mappedAppVersion.dockerInfo?.image,
                    dockerCredentialsId: mappedAppVersion.dockerInfo?.dockerCredentialsId
                ])
            }

            if (mappedAppVersion.resources?.cpu || mappedAppVersion.resources?.disk || mappedAppVersion.resources?.memory) {
                newAppConfig.jobVersion.put('resources', mappedAppVersion.resources)
            }


            if (mappedAppVersion.exposedPorts) {
                newAppConfig.jobVersion.put('exposedPorts', mappedAppVersion.exposedPorts)
            }


        }
        return newAppConfig
    }

    def static mapAppWithoutMail(App app, String projectId) {
        def mappedApp = app.toMap()
        def newAppConfig = [:]

        def technology = app.technology


        newAppConfig.put('job', [
            name          : mappedApp?.name,
            isScheduled   : mappedApp?.isScheduled,
            category      : mappedApp?.category,
            isStreaming   : mappedApp?.isStreaming,
            description   : mappedApp?.description,
            technology    : [id: technology],
            storageSizeInMB: mappedApp?.storageSizeInMB
        ])

        if (projectId) {
            newAppConfig.job.put("projectId", projectId)
        }

        if (mappedApp.id) {
            newAppConfig.job.put('id', mappedApp.id)
        }

        if (mappedApp.alerting?.emails) {
            newAppConfig.app.put('alerting', [
                emails    : mappedApp.alerting?.emails,
                statusList: mappedApp.alerting?.statusList,
                logins    : mappedApp.alerting?.logins
            ])
        }
        return newAppConfig
    }
}

```
## How?

- [ ] Add `ExportApp[] exportApps = []` method `exportArtifactsV2` class `SaagieClient.groovy`.
```
        ExportApp[] exportApps = []
        if (configuration.apps.ids) {
            exportApps = getListAppAndAppVersionsFromConfig()
        }
```
- [ ] Add new methode `getListAppAndAppVersionsFromConfig()`
```
      ExportApp[] getListAppAndAppVersionsFromConfig() {
          return getListAppAndAppVersions(this.&getAppAndAppVersionDetailToExport)
       }
```
- [ ] `getAppAndAppVersionDetailToExport` will populate The variable `ExportApp` with information for the API. this variable will be the one we use to export the application detail.
- [ ] Change return `exportArtifactsFromProjectConfiguration` and `exportAllArtifactsProject` to export `exportApp` variable also.
- [ ] Add to method `getAppDetailRequestFromParam` will contains the query `labWebApp` to get an existing application detail
- [ ]  Add `TechnologyApplicationService` class to get a list of existing technologies for application that will use `repositoriesQuery`.
- [ ]  `getAppAndAppVersionDetailToExport` will populate `ExportApp` variable,we will use it later to export the apps into a zip file.
- [ ] Add `ExportApp[] exportAppList = []` to the `FolderGenerator` class, this class is responsible of generating the data that it contains into an actual zip file.
- [ ] Add `generateFolderForApp` method of the `FolderGenerator` that will generate the json file for `App.json` from the data provided.
- [ ] Add `generateAppVersion` method that will be responsible to add the app version part of the json `App.json` file
- [ ] Add unit tests for exporting new artifact App.

### CODE REVIEW

File: src/main/groovy/io/saagie/plugin/dataops/models/AppMapper.groovy [hunk 1]
```diff
@@ -0,0 +1,91 @@
+package io.saagie.plugin.dataops.models
+
+class AppMapper {
+    App job = new App()
+    AppVersionDTO jobVersion = new AppVersionDTO()
+
+    public AppMapper() {
+        this.AppVersion.resources {
```
Comment (medamineziraoui) [inline src/main/groovy/io/saagie/plugin/dataops/models/AppMapper.groovy:8]: Il faut remplacer ces valeurs avec celle de AppVersionDTO @ZouhairBear

File: src/main/groovy/io/saagie/plugin/dataops/clients/SaagieClient.groovy [hunk 2]
```diff
@@ -930,7 +931,7 @@ class SaagieClient {
         ExportVariables[] exportVariables
         (exportVariables, variablesExportedIsEmpty) = getVariableListIfConfigIsDefined(this.&getListVariablesV1FromConfig)
 
-        return export(exportPipelines, exportJobs, exportVariables, listJobsByNameAndIdFromV1, variablesExportedIsEmpty, true)
+        return export(exportPipelines, exportJobs, exportVariables,null,  listJobsByNameAndIdFromV1, variablesExportedIsEmpty, true)
```
Comment (ZouhairBear) [inline src/main/groovy/io/saagie/plugin/dataops/clients/SaagieClient.groovy:934]: can you add a comment why did you add the null value ?

File: src/main/groovy/io/saagie/plugin/dataops/clients/SaagieClient.groovy [hunk 3]
```diff
@@ -1492,6 +1557,21 @@ class SaagieClient {
...
+    ExportApp[] getListAppAndAppVersions(Closure operation) {
+        checkRequiredConfig(
+            !configuration?.apps?.ids ||
+                !configuration?.exportArtifacts?.export_file
+        )
+        def listAppIdsInt = configuration.apps.ids.collect { it as String }
+        def listAppIds = listAppIdsInt.unique { a, b -> a <=> b }
+        def arrayApps = []
+
+        listAppIds.each { jobId ->
```
Comment (ZouhairBear) [inline src/main/groovy/io/saagie/plugin/dataops/clients/SaagieClient.groovy:1569]: can you refactor the name of `jobId` to `appId` ?

File: src/main/groovy/io/saagie/plugin/dataops/clients/SaagieClient.groovy [hunk 4]
```diff
@@ -1492,6 +1557,21 @@ class SaagieClient {
...
+            arrayApps.add(operation(jobId as String))
```
Comment (ZouhairBear) [inline src/main/groovy/io/saagie/plugin/dataops/clients/SaagieClient.groovy:1570]: same as above

File: src/main/groovy/io/saagie/plugin/dataops/models/AppDTO.groovy [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package io.saagie.plugin.dataops.models
+
+class AppDTO extends JobDTO {
+    int storageSizeInMB;
+
+    void setAppFromApiResult(appDetailResult) {
+        this.storageSizeInMB = appDetailResult.storageSizeInMB;
+        this.setJobFromApiResult(appDetailResult);
```
Comment (ZouhairBear) [inline src/main/groovy/io/saagie/plugin/dataops/models/AppDTO.groovy:8]: well done ! (to not uplicate the code)

File: src/main/groovy/io/saagie/plugin/dataops/models/AppMapper.groovy (hunk 1 above)
Comment (ZouhairBear) [reply src/main/groovy/io/saagie/plugin/dataops/models/AppMapper.groovy:8]: :+1:

File: src/main/groovy/io/saagie/plugin/dataops/utils/SaagieUtils.groovy [hunk 6]
```diff
@@ -352,6 +355,74 @@ class SaagieUtils {
...
+
+        def gqVariables = jsonGenerator.toJson([id: appId])
+
+        def getAppDetailQuery = gq('''
+            query labWebApp($id: UUID!) { labWebApp(id: $id) {    id    name    description    creationDate    isDeletable   storageSizeInMB    instances(limit: 1, checkInPipelineInstance: false) {      id     status     statusDetails     startTime      endTime      version {        number              }          }    versions {     number      creator      creationDate      number      isCurrent      releaseNote      dockerInfo {        image        dockerCredentialsId             }      exposedPorts {        name        port        isAuthenticationRequired       isRewriteUrl        basePathVariableName              }      storagePaths         }    alerting {      emails      statusList      loginEmails {        login        email             }          }    technology {     id  label         }      }}
+        ''', gqVariables)
+        return buildRequestFromQuery(getAppDetailQuery)
+    }
+
+    Request getAppListByNameAndIdRequest(String projectId) {
```
Comment (ZouhairBear) [inline src/main/groovy/io/saagie/plugin/dataops/utils/SaagieUtils.groovy:374]: can you rename it to more significant name ? you don't use the name as a parameter, you only use the `projectId`

File: src/main/groovy/io/saagie/plugin/dataops/utils/SaagieUtils.groovy [hunk 7]
```diff
@@ -352,6 +355,74 @@ class SaagieUtils {
...
+
+        def getUpdateAppVersionQuery = gq('''
+            mutation addAppVersionMutation($appId: UUID!, $appVersion: JobVersionInput!) {  addJobVersion(jobId: $appId, jobVersion: $appVersion) {    number      }}
+        ''', gqVariables)
+        return buildRequestFromQuery(getUpdateAppVersionQuery)
+    }
+
+    Request getProjectCreateAppRequest(App app, AppVersionDTO appVersionDTO) {
+        Map mapedAppAndAppVersion = AppMapper.mapAppAndAppVersionWithoutMail(app, appVersionDTO, configuration.project.id)
+        logger.debug('Updating application version for application with id', appId)
```
Comment (ZouhairBear) [inline src/main/groovy/io/saagie/plugin/dataops/utils/SaagieUtils.groovy:421]: the comment does'nt go with what the mehod do , can you change it ?

File: src/main/groovy/io/saagie/plugin/dataops/utils/directory/FolderGenerator.groovy [hunk 8]
```diff
@@ -41,6 +44,189 @@ class FolderGenerator {
         this.overwrite = overwrite
     }
 
+    void generateFolderFromParams(variablesExportedIsEmpty) {
+        if (variablesExportedIsEmpty && checkExistenceOfJobsPipelinesVariablesAndApps()) {
+            throw new GradleException("Cannot generate zip file")
+        }
+
+        if (checkExistenceOfJobsPipelinesVariablesAndApps()) {
+            throw new GradleException("jobs, pipelines,variables and apps to be exported cannot be empty at the same time, and cannot generate zip file")
```
Comment (ZouhairBear) [inline src/main/groovy/io/saagie/plugin/dataops/utils/directory/FolderGenerator.groovy:53]: we should validate this error message with the client.what do you think ?

File: src/main/groovy/io/saagie/plugin/dataops/utils/directory/FolderGenerator.groovy [hunk 9]
```diff
@@ -41,6 +44,189 @@ class FolderGenerator {
...
+                urlToDownload,
+                urlFileOflocalPackage,
+                client,
+                SaagieUtils.getFileNameFromUrl(downloadUrl)
+            )
+        } catch (IOException e) {
+            throw new GradleException(e.message)
+        }
+    }
+    // TODO add logic for generateFolderForApp
```
Comment (ZouhairBear) [inline src/main/groovy/io/saagie/plugin/dataops/utils/directory/FolderGenerator.groovy:123]: can you remove this TODO ?i think that it's not necessary
